django/db/models/functions/text.py

```diff
@@ -204,6 +207,117 @@ def as_sqlite(self, compiler, connection, **extra_context):
         return super().as_sql(compiler, connection, function='UNICODE', **extra_context)
 
 
+class RegexpFlagMixin:
+    default_flags = {
+        'mariadb': 'c',  # Force case-sensitive matching.
+        'postgresql': 'p',  # Force partial newline-sensitive matching.
+    }
+    inline_flags = {}
+    wrap_pattern = {}
+
+    @cached_property
+    def flag_mapping(self):
+        return {k: str.maketrans(v) for k, v in {
+            'mariadb': {'c': '-i'},
+            'oracle': {'s': 'n'},
+            'postgresql': {'m': 'n'},
+        }.items()}
+
+    def as_sql(self, compiler, connection, **extra_context):
+        expression, pattern, *extra, flags = self.get_source_expressions()
+
+        if connection.vendor == 'mysql' and connection.mysql_is_mariadb:
+            vendor = 'mariadb'
+        else:
+            vendor = connection.vendor
+
+        # Add default flags to ensure consistency between database backends.
+        if vendor in self.default_flags:
+            flags.value = self.default_flags[vendor] + flags.value
+
+        # Wrap pattern in group if required and increment backreferences.
+        if pattern.value and vendor in self.wrap_pattern and not extra_context.get('no_wrap'):
+            pattern.value = '(%s)' % re.sub(
+                r'\\([0-9]+)',
+                lambda m: r'\%d' % (int(m.group(1)) + 1),
+                pattern.value,
+            )
+
+        # Remove duplicate flags preserving the last occurrence of each to
+        # ensures that the last flag is preferred if any are contradictory.
+        # PY38: list() is no longer required to use reversed() on dict().
+        if flags and flags.value:
+            flags.value = ''.join(reversed(list(dict.fromkeys(reversed(flags.value)))))
+
+        # Resolve case-sensitivity flags preferring the last flag specified.
+        if {'c', 'i'}.issubset(flags.value):
+            remove = 'i' if flags.value.index('c') > flags.value.index('i') else 'c'
+            flags.value = flags.value.replace(remove, '')
+
+        if vendor in self.flag_mapping:
+            flags.value = flags.value.translate(self.flag_mapping[vendor])
+
+        if vendor in {'mysql', 'oracle'}:
+            position = Value(1)
+            occurrence = Value(0 if 'g' in flags.value else 1)
+            flags.value = flags.value.replace('g', '')
+            extra += (position, occurrence, flags)
+        elif vendor in self.inline_flags:
+            if flags.value and pattern.value:
+                # FIXME: Can this hack be improved?
+                if '-i' in flags.value:
+                    pattern.value = '(?-i)' + pattern.value
+                    flags.value = flags.value.replace('-i', '')
```

Comment:
This is a bit clunky. Ideas for a cleaner way to handle this with respect to the above code would be appreciated.